testing.rs

```diff
@@ -0,0 +1,621 @@
+/// Inserts a suite of ink! unit tests intended for a contract implementing PSP22 trait.
+/// `$contract` argument should be the name of the contract struct.
+/// `$constructor` argument should be the name of a function, which initializes `$contract`
+/// with the given total supply of tokens.
+/// This macro should be invoked inside `#[ink::contract]` module.
+#[macro_export]
+macro_rules! tests {
+    ($contract:ident, $constructor:path) => {
+        use super::*;
+        use ink::env::{test::*, DefaultEnvironment as E};
+
+        type Event = <$contract as ::ink::reflect::ContractEventBase>::Type;
+
+        // Gathers all emitted events, skip `shift` first, decode the rest and return as vector
+        fn decode_events(shift: usize) -> Vec<Event> {
+            recorded_events()
+                .skip(shift)
+                .map(|e| <Event as scale::Decode>::decode(&mut &e.data[..]).unwrap())
```

Comment:
two options here:
 - either you use `scale::` imports in whole macro and then you should add explicit requirement that `parity-scale-codec` crate should be available in the target crate behind an alias `scale`
 - or you use `::scale::` imports in whole macro and then you should add explicit requirement that `parity-scale-codec` must be a dependency of the target crate **renamed** to `scale`

Reply:
and this applies to every other dependency you assume in this macro to be available (like `ink`)

Reply:
there is also a third approach that _this_ crate exports `scale`, and you use `::psp22::scale::` imports (requiring that the target crate doesn't rename this crate)

Reply:
I think the requirement of having the macro invocation inside the main contract's module silently enforces that both `ink` and `scale` are visible

Reply:
if we put it within this module, then okay, I'm not insisting